core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -463,6 +469,53 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
       }
     }
 
+  def getCatalog(workspaceName: WorkspaceName): Future[PerRequestMessage] =
+    dataSource.inTransaction { dataAccess =>
+      withWorkspaceContext(workspaceName, dataAccess) { workspaceContext =>
+        dataAccess.workspaceQuery.listEmailsAndCatalog(workspaceContext).map { accessResults =>
+          RequestComplete(StatusCodes.OK, accessResults)
+        }
+      }
+    }
+
+  def updateCatalog(workspaceName: WorkspaceName, input: Seq[WorkspaceCatalog]): Future[PerRequestMessage] = {
+
+    val updatesFuture = dataSource.inTransaction { dataAccess =>
+      withWorkspaceContext(workspaceName, dataAccess) { workspaceContext =>
+        determineNewCatalogPermissions(input, dataAccess, workspaceContext)
+      }
+    }
+
+    def updateWorkspaceCatalogPermissions(actualChangesToMake: Map[Either[RawlsUserRef, RawlsGroupRef], Boolean]) = {
+      val (usersToAdd, usersToRemove) = actualChangesToMake.collect { case (Left(userRef), catalog) => userRef -> catalog }.toSeq
+        .partition { case (_, catalog) => catalog }
+      val (groupsToAdd, groupsToRemove) = actualChangesToMake.collect { case (Right(groupRef), catalog) => groupRef -> catalog }.toSeq
+        .partition { case (_, catalog) => catalog }
+
+      dataSource.inTransaction { dataAccess =>
+        withWorkspaceContext(workspaceName, dataAccess) { workspaceContext =>
+          dataAccess.workspaceQuery.insertUserCatalogPermissions(workspaceContext.workspaceId, usersToAdd.map { case (userRef, _) => userRef }) andThen
+            dataAccess.workspaceQuery.deleteUserCatalogPermissions(workspaceContext.workspaceId, usersToRemove.map { case (userRef, _) => userRef }) andThen
+            dataAccess.workspaceQuery.insertGroupCatalogPermissions(workspaceContext.workspaceId, groupsToAdd.map { case (groupRef, _) => groupRef }) andThen
+            dataAccess.workspaceQuery.deleteGroupCatalogPermissions(workspaceContext.workspaceId, groupsToRemove.map { case (groupRef, _) => groupRef })
+        }
+      }
+    }
+
+    for {
+      (emailsNotFound, actualChangesToMake) <- updatesFuture
+      savedPermissions <- updateWorkspaceCatalogPermissions(actualChangesToMake)
```

Comment:
Good catch.  Another reason to do this all in one transaction is atomicity / concurrency protection: two threads might stomp on each other's results if they can interleave these two steps in different transactions.